tools/wake-format/README.md

```diff
@@ -0,0 +1,481 @@
+
+# Usage
+
+The standard usage of `wake-format` is to format a set of wake files in place.
+
+`wake-format -i source.wake source2.wake source3.wake`
+
+`wake-format` supports other flows for various use cases. The `-n`/`--dry-run` flag is ideal for CI/CD to check if the source needs to be formatted. It exits with a non-zero status code when a file would be changed by `wake-format`
+
+Running `wake-format` without `-i` will emit the file to `stdout` instead of overwritting it.
+
+# Documentation
+
+The information below should be helpful to contribute to `wake-format`
+
+## Building
+
+`wake-format` is built in the same manner as `wake`. Running `make` or `./bin/wake build default` from within the wake dir will produce the `wake-format` binary.
+
+The locally build `wake-format` can be found at `./bin/wake-format.native-cpp14-release`
+
+## wake-format Fluent API/DSL
+
+`wake-format` relies heavily on a custom fluent API/DSL to generate the formatting rules. Usage of the API is documented below. See [here](TODO) for design decisions, invariants, and engineering tradeoffs considered.
+
+### Usage
+
+All formatting rules start by creating a basic formatting object using `fmt()`. From there functions are added to describe the expectations of the current [CST node](TODO) and to return the appropiate text.
+
+```c++
+auto basic_fmt = fmt()
+    .space()
+    .newline();
+```
+
+creates a formatter that returns a space and a newline.
+
+Once a formatter has been defined, the `format()` function is used to run the formatter and return the formated source. The source is returned as a [wcl::doc](TODO) for efficient concatenation and geometry lookup. `format()` takes a `ctx_t` and a `CSTElement`. The `ctx_t` (context) tracks the context (such as nest level) that all subnodes should be formatted in, and the `CSTElement` is the current node being formatted.
+
+```c++
+ctx_t ctx;
+CSTElement node = ""; // An empty node
+
+wcl::doc doc = fmt()
+  .space()
+  .newline()
+  .format(ctx, node);
+
+std::string s = doc.as_string(); // " \n"
+```
+
+`format()` expects to consume the entire `CSTElement` node and will `assert()` if that is not the case. Asserting may seem heavily handed, however this is a very powerful method for identifying and fixing nodes with unexpected contents. 
+
+### API Functions & Examples
+
+**CSTElement Manipulation**
+
+Manipulation of the current state of the formatter relating to CSTElement nodes.
+
+**`token(uint8_t id)`**
+
+`assert()`s the current node.id() is `id` then emits the string representation of the node.
+
+```c++
+CSTElement node = "def";
+auto doc = fmt()
+  .token(TOKEN_KW_DEF)
+  .space()
+  .format(ctx, node);
+doc.as_string(); // "def "
+```
+
+**`token(uint8_t id, const char* str)`**
+
+`assert()`s the current node.id() is `id` then emits `str`.
+
+```c++
+CSTElement node = "def";
+auto doc = fmt()
+  .token(TOKEN_KW_DEF, "foo")
+  .space()
+  .format(ctx, node);
+doc.as_string(); // "foo "
```

Comment:
is `"foo "` intended to have the trailing space char? when I look at `next()` the `id` seems to be separated from spaces around it

Reply:
The trailing space comes from `space()` on line 84. I think I was trying to make the example slightly less trivial however if it is confusing it should just be removed

Reply:
oh somehow my eyes entirely slid over the `space()`